### DISCUSSION TITLE
Lighthouse positioning for F1/10 vehicles

### DISCUSSION THREAD

**Discussion by flochkristof** [bitcraze/discussions#218] [Q&A] [answered]
Hi,

we are planning on switching to lighthouse positioning from motion capture in our lab. For the crazyflie drones the transition seem to be straight forward. However, we also use autonomous [f1tenth](https://f1tenth.org/build.html) cars equipped with an Nvidia Jetson TX2 computer board. We would like to connect the lighthouse deck to this board as it has UART and I2C communication ports.

In your opinion what is the easiest way to retrieve the position data from the deck? Which part(s) of the crazyflie firmware could we use or do you have any sample code snippets that would help?

Thanks in advance!

**Comment by krichardsson** [comment, marked as answer]
Hi, 
this sounds like an exciting project!

What you want to do will require some work, but I think it will be doable. I will try to outline what the code is doing, to help you understand what you need to do if you want to connect the lighthouse deck to some other device than a Crazyflie.
1. **Initialize.** The process starts in [the lighthouse deck driver](https://github.com/bitcraze/crazyflie-firmware/blob/2022.05/src/deck/drivers/src/lighthouse.c) which calls [`lighthouseCoreInit()`](https://github.com/bitcraze/crazyflie-firmware/blob/8c606d1f8f89f1f00462dfa1f294a431a321242d/src/modules/src/lighthouse/lighthouse_core.c#L139-L146). This will get data from the persistent storage and initialize the memory mapping system that is used to read/write data from a PC. 
2. **Boot the deck.**  The deck is booted as a part of the start up of the [lighthouse task](https://github.com/bitcraze/crazyflie-firmware/blob/8c606d1f8f89f1f00462dfa1f294a431a321242d/src/modules/src/lighthouse/lighthouse_core.c#L456-L512). The call to [`lighthouseDeckFlasherCheckVersionAndBoot()`](https://github.com/bitcraze/crazyflie-firmware/blob/8c606d1f8f89f1f00462dfa1f294a431a321242d/src/modules/src/lighthouse/lighthouse_core.c#L466) boots the deck via i2c. 
3. **Run the main loop** The main action is in [the loop where data is read from the deck through the UART and is pushed to the pulse processor](https://github.com/bitcraze/crazyflie-firmware/blob/8c606d1f8f89f1f00462dfa1f294a431a321242d/src/modules/src/lighthouse/lighthouse_core.c#L486-L503) for [further crunching](https://github.com/bitcraze/crazyflie-firmware/blob/8c606d1f8f89f1f00462dfa1f294a431a321242d/src/modules/src/lighthouse/lighthouse_core.c#L392).
4. **Interpretation of data** The purpose of the [pulse processor ](https://github.com/bitcraze/crazyflie-firmware/blob/2022.05/src/utils/src/lighthouse/pulse_processor.c)is to interpret the data from the deck and convert it to sweep angles. This is also where calibration data is applied. Our code supports both LH1 and LH2 which you probably don't need, so you can probably simplify things if you want to.
5. **Initial position estimation** In [the lighthouse_position_est.c](https://github.com/bitcraze/crazyflie-firmware/blob/2022.05/src/modules/src/lighthouse/lighthouse_position_est.c) file, the angles are pushed on to the estimator. If the crossing beam method is used, the position is computed here. The yaw estimation is also done here, it is a bit of a hack and should be moved completely to the estimator instead. Results from computations in this file are pushed to the estimator (search for "enqueue").  
6. **Sensor fusion** In the kalman estimator, the data from step 5 is fused with other sensor data for a final state estimate.

Files that are located in `src/modules/src/lighthouse` have dependencies to freeRTOS while files in `src/utils/src/lighthouse` should be more independent and easier to reuse. 

You can choose how much of this process you want to reuse, but you will probably need most of steps 1-4 to get any useful data (sweep angles). After that you can chose how to process them, depending on your setup.

Another option is to use a Crazyflie (or Bolt) on your platform as an intermediate step and extract data at an appropriate level for further processing in your platform.

**Reply by flochkristof** [reply]
Thanks for the fast and detailed response!

**Reply by krichardsson** [reply]
You're welcome! 
It would be interesting to hear if it worked out later on :-)
